eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,322 @@
+	if err != nil {
+		logger.Errorw("failed to parse request body", zap.Error(err))
+		common.SendErrorResponse(writer, err.Error())
+		route.Metrics.EventProcessingFailed(route.EventSourceName, route.EventName)
+		return
+	}
+
+	event := &events.BitbucketServerEventData{
+		Headers:  request.Header,
+		Body:     (*json.RawMessage)(&body),
+		Metadata: router.bitbucketserverEventSource.Metadata,
+	}
+
+	eventBody, err := json.Marshal(event)
+	if err != nil {
+		logger.Info("failed to marshal event")
+		common.SendErrorResponse(writer, "invalid event")
+		route.Metrics.EventProcessingFailed(route.EventSourceName, route.EventName)
+		return
+	}
+
+	logger.Info("dispatching event on route's data channel")
+	route.DataCh <- eventBody
+
+	logger.Info("request successfully processed")
+	common.SendSuccessResponse(writer, "success")
+}
+
+// PostActivate performs operations once the route is activated and ready to consume requests
+func (router *Router) PostActivate() error {
+	route := router.GetRoute()
+	bitbucketserverEventSource := router.bitbucketserverEventSource
+
+	logger := route.Logger.With(
+		logging.LabelEndpoint, route.Context.Endpoint,
+		logging.LabelPort, route.Context.Port,
+		logging.LabelHTTPMethod, route.Context.Method,
+		"project-key", bitbucketserverEventSource.ProjectKey,
+		"repository-slug", bitbucketserverEventSource.RepositorySlug,
+		"base-url", bitbucketserverEventSource.BitbucketServerBaseURL,
+	)
+
+	logger.Info("retrieving the access token credentials...")
+	bitbucketCredentials, err := router.getCredentials(bitbucketserverEventSource.AccessToken)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver credentials. err: %+v", err)
+	}
+
+	logger.Info("setting up the client to connect to Bitbucket Server...")
+	bitbucketConfig := bitbucketv1.NewConfiguration(bitbucketserverEventSource.BitbucketServerBaseURL)
+	bitbucketConfig.AddDefaultHeader("x-atlassian-token", "no-check")
+	bitbucketConfig.AddDefaultHeader("x-requested-with", "XMLHttpRequest")
+
+	ctx := context.WithValue(context.Background(), bitbucketv1.ContextAccessToken, bitbucketCredentials.token)
+
+	// When running multiple replicas of the eventsource, they will all try to create the webhook.
+	// Randomly sleep some time to mitigate the issue.
+	s1 := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(s1)
+	time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
+	err = router.CreateBitbucketWebhook(ctx, bitbucketConfig)
+
+	if err != nil {
+		logger.Errorw("failed to create Bitbucket webhook", zap.Error(err))
+	}
+
+	go func() {
+		// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
+		// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from bitbucket.
+		// This is a workround to mitigate the race conditions.
+		logger.Info("starting bitbucket hooks manager daemon")
+		ticker := time.NewTicker(60 * time.Second)
+		defer ticker.Stop()
+		for {
```

Comment:
Putting this logic in `PostActive()` has a problem that `<- ctx.Done()` below never gets a chance to reach, try it by tailing the pod log and kill it to see if `exiting bitbucket hooks manager daemon` is printed. Correct me if I'm wrong.

Reply:
1. Why does the goroutine keep on creating webhooks? shouldn't it return after the first successful bitbucket webhook creation? 
2. Shouldn't the function return after line 164 if err == nil.
3. As @whynowy  mentioned, I am not sure if retrying webhook creation every 1 min makes sense. We can instead use exponential backoff and try to create the webhook as opposed to running a daemon.
4. Also if we go with the daemon, we should use signal.notify on sigterm instead of ctx.Done on line 179

Reply:
Thanks for the review. I'll clean this up. The current PR is a mix of the gitlab eventsource (that suffers from the same race conditions) and the fixes that have been done on the github eventsource. 

Like @whynowy mentioned before: I don't like the way you need to cope with these kind of race conditions in every eventsource that creates external resources.

There's obviously a demand for this use case (the Bitbucket eventsource is the nr1 voted issue currently), but I think a decision has to be made on how to implement this in a generic way.

I'm not that familiar with the design of Argo Events, but maybe this can be solved using a install/uninstall Job when creating/deleting the EventSource k8s resource, so it's not tied to the lifecycle of the individual Pods. Or maybe using a finalizer to clean up the external resource. 

This way the responsibility is moved to the eventsource-controller and the external resources can be left alone during RollingUpdates.

Any thoughts?

Reply:
Is `DeleteHookOnFinish` a must have feature? It is root cause of the race condition. If we abandon it, it avoids lots of problem. Question is how many ppl want it, I don't know.

My suggestion:
1. No `DeleteHookOnFinish` support.
2. If ppl do want `DeleteHookOnFinish`, use the workaround at this time.

Reply:
If you want to go "full gitops", the `DeleteHookOnFinish` is a must have feature. It shouldn't even be possible to *not* remove the webhook, because it doesn't make sense to keep the webhook around when there's no EventSource to process the events.

I think this feature is very powerful and in fact, the video of @vfarcic linked from the frontpage is what has drawn me towards the whole Argo "suite" of tools.

The point is not that it's *hard* to configure or cleanup the webhooks in Bitbucket (or Github, or Gitlab, etc.), the point is that it's a *manual* step I'd like to avoid as much as possible.

I think there's a big opportunity for Argo Events here, because there are a lot of systems out there that support some kind of webhooks that can be controlled using an API.

For now, I'll just fix the Bitbucket Server EventSource to handle the race conditions a bit better and keep it on par with the Github implementation.

Thanks!

Reply:
Sounds good, thanks!